#edgeR_DEG_analysis

#We performed time-course mRNA-seq using zebrafish whole brains, total 125 samples.

#read counts matrix was generated by HISAT2-Stringtie 
#reference genome: Danio_rerio.GRCz11.dna.primary_assembly.fa
#reference annotaion: Danio_rerio.GRCz11.107.chr.gtf.gz 


# sourse packages and functions -------------------------------------------

source("code/packages_and functions.R")


# working dir and session info --------------------------------------------


getwd()
#print the session and version info
sessionInfo()
rstudioapi::versionInfo()

#save them to file
writeLines(capture.output(sessionInfo()), "code/sessionInfo.txt")
writeLines(capture.output(rstudioapi::versionInfo()), "code/versionInfo.txt")


# load read count and gene, sample information matrix --------------------------------------------------------

#read.matrix
countData <- as.matrix(read.csv("./data/star_wt_gene_count_matrix.csv",row.names = 1))

#sample info.
colData <- read.csv("./data/sample_info.csv", row.names = 1)

#col.check and name.match
all(colData$samples %in% colnames(countData))

class(countData)
dim(countData)

#gene.ID.
ID.gene = str_split(string = rownames(countData),pattern = "\\|")
ID.and.gene=do.call(rbind.data.frame, ID.gene)
ID.gene=data.frame(rownames(countData),ID.and.gene[,c(1,2)])
colnames(ID.gene)=c("id","ensembl_gene_id", "zfin_id_symbol")


#adding Ensembl annotation
library(biomaRt)
listMarts()
ensembl = useEnsembl(biomart="ensembl", mirror = "www") #if not working, try with "useast" or"uswest"
datasets <- listDatasets(ensembl)
head(datasets)
ensembl = useDataset("drerio_gene_ensembl",mart=ensembl)
filters = listFilters(ensembl)
attributes = listAttributes(ensembl)

ft_txid.anno=getBM(attributes =c("ensembl_gene_id","description"),
                   filters = data.frame("ensembl_gene_id" = ID.gene$ensembl_gene_id), mart = ensembl)

#ft_txid.anno.uq= ft_txid.anno[- c(which(duplicated(ft_txid.anno$zfin_id_id))),]
#ft_txid.anno.dup = ft_txid.anno[c(which(duplicated(ft_txid.anno$zfin_id_id))),]

library(plyr)
ID.gene = join(ID.gene,ft_txid.anno, by="ensembl_gene_id")
d= duplicated(ID.gene$id)
row.names(ID.gene)=ID.gene$id

detach("package:plyr", unload = TRUE)

library(dplyr)



# DEGList -----------------------------------------------------------------

#making a DEGList
x = DGEList(counts = countData,samples = colData$samples)
samples = colData$samples

#adding sample information to DEGList
group = colData$exp_cond
x$samples$group = as.factor(colData$exp_cond)
lane = colData$genotype
x$samples$lane = as.factor(colData$genotype)
time = colData$time
x$samples$time = as.factor(colData$time)
identity = colData$identity
x$samples$identity = as.factor(colData$identity)


ft_x= x[c(which(rownames(x) %in% ID.gene$id)),]
ft_x=ft_x[rownames(ID.gene),]
ft_x$gene = ID.gene
ft_x


#Filtering_genes with very low counts by group --------------------------------------------

keep.exprs <- filterByExpr(ft_x, group=identity)
ft_x.2 <- ft_x[keep.exprs,, keep.lib.sizes=FALSE]
dim(ft_x.2)

#normalization 
ft_x.2.norm <- calcNormFactors(ft_x.2, method = "TMM")
#write.csv(ft_x.2.norm$counts, "./outputs/ft_x.2.norm.count.csv")
#write.csv(ft_x.2.norm$gene, "./outputs/ft_x.2.norm.gene.csv")
#write.csv(ft_x.2.norm$samples, "./outputs/ft_x.2.norm.sample.csv")


# pMDS plot ---------------------------------------------------------------


col = as.numeric(ft_x.2.norm$samples$identity)
samp = as.numeric(ft_x.2.norm$samples$lane)

pMDs=plotMDS(ft_x.2.norm, col=col, pch = samp, cex = 1, gene.selection="pairwise")
toplot <- data.frame(Dim1 = pMDs$x, Dim2 = pMDs$y, 
                     Group = factor(ft_x.2.norm$samples$time, levels = c("d6","d13","d120","LD")), samp =ft_x.2.norm$samples$lane )
library(ggforce)

pdf(file = "./figures/MDS.plot.star.pdf", 
    width = 8, # The width of the plot in inches
    height = 5)   # The directory you want to save the file in

p = ggplot(toplot, aes(Dim1, Dim2, shape = samp)) +
  geom_point(aes(colour = Group, size = 2))+
  xlim(-5, 3.1)+
  ylim(-2,2.5)+
  scale_color_viridis(discrete = T, alpha = 1)+
  labs(
    #title = "MDS plot: wild type vs. star:bPAC ",
    #subtitle = "",
    #caption = "Data from the 1974 Motor Trend US magazine.",
    #tag = "Figure 1",
    x = "Leading logFC dim 1",
    y = "Leading logFC dim 2",
    colour = "time points",
    shape = "genotype"
  )
p+theme_light()+ guides(size = guide_legend()) +theme(text = element_text(size=15))


dev.off()



# design matrix for comparison --------------------------------------------

#matrix design
design = model.matrix(~0+identity)
colnames(design) = c(gsub("identity","",colnames(design)))
rownames(design) = colData$samples
#
ft_x.2.norm <- estimateDisp(ft_x.2.norm,design)
ft_x.2.norm$common.dispersion
#countr.matrix

countr.matrix = makeContrasts(
  d6_s.NEGvsWT = star_neg_d6 - tu_wt_d6,
  d13_s.NEGvsWT = star_neg_d13 - tu_wt_d13,
  d120_s.NEGvsWT = star_neg_d120 - tu_wt_d120,
  d120_LD_s.NEGvsWT = star_neg_LD- tu_wt_LD,
  d6_s.POSvsWT = star_pos_d6 - tu_wt_d6,
  d13_s.POSvsWT = star_pos_d13 - tu_wt_d13,
  d120_s.POSvsWT = star_pos_d120 - tu_wt_d120,
  d120_LD_s.POSvsWT = star_pos_LD - tu_wt_LD,
  levels = colnames(design))
countr.matrix


# Testing for DEG ---------------------------------------------------------

gfit <- glmQLFit(ft_x.2.norm, design, robust = T)
c.gfit = glmQLFTest(gfit, contrast=countr.matrix)

call.name = colnames(countr.matrix)

for(i in c(1:8)){
  qn = paste0("qlf",i)  
  assign(qn, glmQLFTest(gfit, contrast=countr.matrix[,call.name[i]]))
  
  print(i)
}

#test (FDR)
for(i in c(1:8)){
  qlf = get(paste0("qlf",i))
  FDR = paste0("FDR",i)
  assign(FDR, p.adjust(qlf$table$PValue, method="BH"))
  
  print(i)
}


# fit.modeling ------------------------------------------------------------

pdf(file = "./figures/fit.modeling.pdf",   # The directory you want to save the file in
    width = 11.69, # The width of the plot in inches
    height = 8.27) # The height of the plot in inches
par(mfrow=c(2,2))
v <- voom(ft_x.2.norm, design, plot=TRUE)
v <- voom(gfit$fitted.values, design, plot=TRUE)
plotBCV(ft_x.2.norm)
plotQLDisp(gfit)
dev.off()


# combing tables_ logFC+FDR+annotations+CPM ------------------------------------------

#annotation
gfit.ens = ID.gene[c(which(ID.gene$id %in% rownames(gfit))),]

# count per million (CPM)
cmp_all = cpm(ft_x.2.norm)

#log2FC with FDR
for(i in 1:8){
  qn = paste0("qlf",i)
  qt = topTags(get(qn),n = Inf)
  assign(paste0("topqlf",i), data.frame(qt$table))
}
#match row order
for(i in 1:8){
  qn = get(paste0("topqlf",i))
  qn = qn[gfit.ens$id,]
  assign(paste0("topqlf",i), qn)
}

### design
'
1: d6_s.NEGvsWT = star_neg_d6 - tu_wt_d6,         = 4 - 12
2: d13_s.NEGvsWT = star_neg_d13 - tu_wt_d13,      = 3 - 11
3: d120_s.NEGvsWT = star_neg_d120 - tu_wt_d120,         = 2 - 10
4: d120_LD_s.NEGvsWT = star_neg_d120_LD- tu_wt_d120_LD, = 1 - 9
5: d6_s.POSvsWT = star_pos_d6 - tu_wt_d6,         = 8 - 12
6: d13_s.POSvsWT = star_pos_d13 - tu_wt_d13,      = 7 - 11
7: d120_s.POSvsWT = star_pos_d120 - tu_wt_d120,         = 6 - 10
8: d120_LD_s.POSvsWT = star_pos_d120_LD - tu_wt_d120_LD, 5 - 9

 "star_neg_LD"   "star_neg_d120" "star_neg_d13"  "star_neg_d6"   
 "star_pos_LD"   "star_pos_d120" "star_pos_d13" "star_pos_d6"   
 "tu_wt_LD"      "tu_wt_d120"    "tu_wt_d13"     "tu_wt_d6"  
'
sample.order = unique(colData$identity)
sample.order
mt=c(1)
for(i in 1:12){
  mt = c(mt, (5*i)+1)
}

comp.matrix = data.frame("f"=c(4,3,2,1,8,7,6,5),
                         "r"= c(12,11,10,9,12,11,10,9))

#fc+counts for bPAC+ vs. WT; deg.cpm1~10
for(i in 1:8){
  qlf = get(paste0("topqlf",i))
  a=mt[comp.matrix[i,1]]
  b=mt[comp.matrix[i,2]]
  qlfb = cbind(qlf,gfit.ens,cmp_all[,c(a:(a+4),b:(b+4))])
  write.csv(qlfb, paste0("./data/topqlf.cpm.",i,".csv"))
  assign(paste0("deg.cpm",i),qlfb)
  print(i)
}



# visualization, volcano plot and MDplot -----------------------------------


pdf(file = "./figures/VC.MDplot.pos-wt.pdf",   # The directory you want to save the file in
    width = 8.27, # The width of the plot in inches
    height = 11.69) # The height of the plot in inches

par(mfrow=c(2,1))

#deg.cpm,tx
for(i in c(1:8)){
  
  deg.cpm = get(paste0("deg.cpm",i))
  
  fc_sig_results_txs=which((deg.cpm$logFC)>=log2(2))
  fc_sig_results_tr_plot=deg.cpm[fc_sig_results_txs,] 
  
  fc_sig_results_txs4=which(deg.cpm$logFC>=log2(2))
  fc_sig_results_tr_plot4=deg.cpm[fc_sig_results_txs4,]
  fc_sig_results_txs5=which(deg.cpm$logFC <= -log2(2))
  fc_sig_results_tr_plot5=deg.cpm[fc_sig_results_txs5,]
  
  fc_sig_results_txs2=which(abs(deg.cpm$logFC)>=log2(4))
  fc_sig_results_tr_plot2=deg.cpm[fc_sig_results_txs2,]
  
  fc_sig_results_txs6=which(deg.cpm$logFC>= log2(4))
  fc_sig_results_tr_plot6=deg.cpm[fc_sig_results_txs6,]
  fc_sig_results_txs7=which(deg.cpm$logFC<= -log2(4))
  fc_sig_results_tr_plot7=deg.cpm[fc_sig_results_txs7,]
  
  fc_sig_results_txs3=which(abs(deg.cpm$FDR)>=0.01)
  fc_sig_results_tr_plot3=deg.cpm[fc_sig_results_txs3,]
  
  plot(deg.cpm$logFC, -log10(deg.cpm$FDR), main=c(call.name[i],'tx_qval<0.01'), 
       xlab = "log2fc",ylab="-log10(q-value)", ylim=range(0:15) ,col='grey',cex= 0.5, pch=1)
  
  points(fc_sig_results_tr_plot4$logFC,-log10(fc_sig_results_tr_plot4$FDR), col='pink', cex=0.5,pch=16)
  points(fc_sig_results_tr_plot5$logFC,-log10(fc_sig_results_tr_plot5$FDR), col='lightblue', cex=0.5,pch=16)
  points(fc_sig_results_tr_plot6$logFC,-log10(fc_sig_results_tr_plot6$FDR), col='red', cex=0.5,pch=16)
  points(fc_sig_results_tr_plot7$logFC,-log10(fc_sig_results_tr_plot7$FDR), col='blue', cex=0.5,pch=16)
  points(fc_sig_results_tr_plot3$logFC,-log10(fc_sig_results_tr_plot3$FDR), col='grey',cex=0.6, pch=16)
  
  
  abline(h=-log10(0.01), col='black',lwd=1)
  abline(v=c(log2(4),-log2(4)),lty =2, col='grey', lwd=1)
  abline(v=c(log2(2),-log2(2)),lty =2, col='grey', lwd=1)
  
  up2fc=nrow(filter(fc_sig_results_tr_plot4, FDR <=0.01)) 
  up4fc=nrow(filter(fc_sig_results_tr_plot6, FDR <=0.01)) 
  down2fc=nrow(filter(fc_sig_results_tr_plot5, FDR <=0.01)) 
  down4fc=nrow(filter(fc_sig_results_tr_plot7, FDR <=0.01)) 
  total_sig=nrow(filter(deg.cpm, FDR <=0.01)) 
  
  legend('topright',c('Up > 4fc,',up4fc,'Up > 2fc,',up2fc), ncol=2,bty='n',pch = c(19,NA,19,NA), col=c('red',NA,'pink',NA),cex = 0.6)
  legend('topleft',c('Down > 4fc,',down4fc,'Down > 2fc,',down2fc), ncol=2,bty='n',pch = c(19,NA,19,NA), col=c('blue',NA,'lightblue',NA),cex = 0.6)
  legend("bottomright",c('Total q-val < 0.01,',total_sig),bty='n',cex = 0.6)
  
  qlf = get(paste0("qlf",i))
  plotMD(qlf, ylim=(c(-10,10)), cex=0.5)
  abline(h=c(-1,1), col="cyan")
  
}
dev.off()


# rough calculation of the number of DEGs -----------------------------------

sig_table = summary(decideTests(qlf1,p.value = 0.01,lfc = log2(2)))
for(i in c(2:8)){
  qlf = get(paste0("qlf",i))
  sig_table1 = summary(decideTests(qlf,p.value = 0.01,lfc = log2(2)))
  sig_table = cbind (sig_table,sig_table1)
}
write.csv(sig_table, "./outputs/sig.table.csv")


##read homologs of human and mice ----------------------------------------


zh.all = read.csv("./data/zh.all.csv",header = T)
zm.all = read.csv("./data/zm.all.csv",header = T)

# modification of table for DEG Filtering -------------------------------------------------------

time.name = c("d6.sn", "d13.sn", "d120.sn", "LD.sn","d6", "d13", "d120", "LD")


#read.tables and adding mean CPM per group, t.test

for(i in 1:8){
  tb = get(paste0("deg.cpm",i))
  ttb=c()
  mcp=c()
  mcw=c()
  colnames(tb)=c("logFC","logCPM","F","PValue","FDR","id","ensembl_gene_id","zfin_id_symbol","description","pos1","pos2","pos3","pos4","pos5","wt1","wt2","wt3","wt4","wt5")
  tb = tb %>% dplyr::arrange(desc(logCPM))
  d = duplicated(tb$zfin_id_symbol)
  tb=tb[!d,]
  
  for (k in 1:nrow(tb)) {
    tt=t.test(as.matrix(tb[k,c(10:14)]),as.matrix(tb[k,c(15:19)]))
    mcp=c(mcp,tt$estimate[1])
    mcw=c(mcw,tt$estimate[2])
    ttb=c(ttb,tt$p.value)
  }
  tb[,"mean.cpm.pos"]=mcp
  tb[,"mean.cpm.wt"]=mcw
  tb[,"t.test"]=ttb
  tb[,"source"] = time.name[i]
  name = paste0("deg.cpm",i)
  assign(name,tb)
}


# combining DEG tables for 6, 13, 120 dpf and 120 + LD  

gene_all = deg.cpm1
for (i in c(2:8)) {
  deg= get(paste0("deg.cpm",i))
  gene_all=rbind(gene_all,deg)
}

#write.csv(gene_all,"./outputs/genes_all.csv")

# entrezid matching
prot_id_map <- clusterProfiler::bitr(unique(gene_all$zfin_id_symbol), "SYMBOL",
                                     "ENTREZID", "org.Dr.eg.db")
d=duplicated(prot_id_map$SYMBOL)
prot_id_map=prot_id_map[!d,]
rownames(prot_id_map) <- prot_id_map$SYMBOL


# Filtering of DEGs -------------------------------------------------------

# filter out genes, both meanCPM lower than 5, p-val from t.test higher than 0.05, 
# FDR higher than 0.05, do not have ENTREZID   

for(i in 1:8){
  deg=get(paste0("deg.cpm",i))
  deg = deg %>% filter(mean.cpm.pos > 5|mean.cpm.wt >5)%>% filter(t.test < 0.05)%>% 
    filter(FDR < 0.05)%>%filter(zfin_id_symbol %in% prot_id_map$SYMBOL) %>% arrange(desc(logCPM))
  d = duplicated(deg$zfin_id_symbol)
  deg=deg[!d,]
  name = paste0("ft.deg.cpm",i)
  assign(name,deg)
}

#listing of significant DEGs 

b.ft.id = ft.deg.cpm5$id
for (i in c(1:8)) {
  deg= get(paste0("ft.deg.cpm",i))
  b.ft.id=unique(c(b.ft.id,deg$id))
}

#ft_gene_all
ft.gene_all = ft.deg.cpm5
for (i in c(1:8)) {
  deg= get(paste0("ft.deg.cpm",i))
  ft.gene_all=rbind(ft.gene_all,deg)
}

write.csv(ft.gene_all,"./outputs/ft.gene_all.csv")

#seperating DEGs
for (i in c(1:8)) {
  deg=get(paste0("ft.deg.cpm",i))
  deg.up = deg %>% filter(logFC > 1) %>% filter(FDR < 0.01)%>% arrange(desc(logFC))
  deg.dn = deg %>% filter(logFC < -1) %>% filter(FDR < 0.01) %>% arrange(logFC)
  name1 = paste0("ft.deg.cpm.up",i)
  assign(name1,deg.up)
  name2 = paste0("ft.deg.cpm.dn",i)
  assign(name2,deg.dn)
  
}

# adjust of FC threshold to 1.5FC for 120 dpf 

for (i in c(3,7)) {
  deg=get(paste0("ft.deg.cpm",i))
  deg.up = deg %>% filter(logFC > log2(1.5)) %>% filter(FDR < 0.01)%>% arrange(desc(logFC))
  deg.dn = deg %>% filter(logFC < -log2(1.5))  %>% filter(FDR < 0.01)%>% arrange(logFC)
  name1 = paste0("ft.deg.cpm.up",i)
  assign(name1,deg.up)
  name2 = paste0("ft.deg.cpm.dn",i)
  assign(name2,deg.dn)
  
}

# combining tables 
deg_all.up = ft.deg.cpm.up1
for (i in c(2:8)) {
  deg= get(paste0("ft.deg.cpm.up",i))
  deg_all.up=rbind(deg_all.up,deg)
}

deg_all.dn = ft.deg.cpm.dn1
for (i in c(2:8)) {
  deg= get(paste0("ft.deg.cpm.dn",i))
  deg_all.dn=rbind(deg_all.dn,deg)
}

deg.all = rbind(deg_all.up,deg_all.dn)

#s.table1a
write.csv(deg.all[,c(7:8,1,2,4,5,9,20:23)],"./outputs/s.table1a_all.DEGs.csv")

#listing DEGs
deg_all = unique(deg.all$zfin_id_symbol)

#### GC_respopnding genes


GC_path = c("nr3c1","nr3c2","igfbp1a","fkbp5","tsc22d3","hsd11b2")

# gc_path
gcp.deg = gene_all%>% filter(zfin_id_symbol %in% GC_path)%>%arrange(factor(zfin_id_symbol, level = GC_path ))%>% 
  filter(source %in% time.name[c(1,2,3,5,6,7)])
#gcp.deg= gcp.deg[,c("zfin_id_symbol", "logFC", "FDR","source")]
write.csv(gcp.deg, "./outputs/gc_responding_Gene.exp.csv")

GC_pathway = c(rep("nr3c1",4),rep("nr3c2",4),rep("igfbp1a",4),rep("fkbp5",4),rep("tsc22d3",4),rep("hsd11b2",4))
time_point <- rep(c(rep("d120",2),rep("LD",2)),6)
geno = rep(c("wt","bPAC+"),12)



# heatmap and venn diagrams,contains codes for Fig3AB and s.Fig2A:  ---------------------------

source("./code/DEGs_heatmap_venndiagram.R")

# GO enrichment analysis, contains codes for s.table2 a-g -----------------------------------------------------
# for the Fig xx need to select terms or remove redundant terms

source("./code/GO_enrichment.R")

# figure3D_GOs ------------------------------------------------------------

#TOP GO terms were selected by semi-manually to avoid redundancy of the GO terms
##### GO dot plots

sel.go.bp.up = read.csv("./data/selectedGO_up.csv",header = T,fileEncoding="UTF-8-BOM")
sel.go.bp.dn = read.csv("./data/selectedGO_dn.csv",header = T,fileEncoding="UTF-8-BOM")

#up
time = sel.go.bp.up$Cluster
term = sel.go.bp.up$Description
for (i in 1:length(term)) {
  if (nchar(term[i])>30) {
    term[i]=paste0(substr(term[i],1,30),"...")
  } else {term[i]=term[i]}
}
adj.p = sel.go.bp.up$p.adjust
ratio = (sel.go.bp.up$ratio.catg)

dotp = data_frame(time,term,adj.p,ratio)
dotp$time = factor(dotp$time, level = c("d6" , "d13" , "d120"))
dotp$term = factor(dotp$term, level = unique(term))
#17 terms

tiff("./figures/Fig3D_GO.BP_up.10x11.tiff", height = 9, width = 10, res = 1200, compression = "lzw", units = "cm")

gp=ggplot(dotp)+
  geom_point(aes(x = time, y = term, size = ratio*100, col = adj.p))+
  theme_light() +
  theme(text = element_text(size = 10),
        #legend.position = "none",
        #axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        legend.key.size = unit(0.3,"cm"),
        legend.text = element_text(size=8),
        legend.title = element_text(size=9),
        plot.title = element_text(size = 10,face="bold"),
        #strip.text.x = element_text(face = "italic"),
  )+
  scale_colour_gradientn(colours = c("#D73027","#FFFFBF"))+
  labs(title = "Top_up_GO:BP",x="",y="")
plot(gp)
dev.off()

#down
time = sel.go.bp.dn$Cluster
term = sel.go.bp.dn$Description
for (i in 1:length(term)) {
  if (nchar(term[i])>30) {
    term[i]=paste0(substr(term[i],1,30),"...")
  } else {term[i]=term[i]}
}
adj.p = sel.go.bp.dn$p.adjust
ratio = (sel.go.bp.dn$ ratio.catg)

dotp = data_frame(time,term,adj.p,ratio)
dotp$time = factor(dotp$time, level = c("d6" , "d13" , "d120"))
dotp$term = factor(dotp$term, level = unique(term))
#17

tiff("./figures/Fig3D_GO.BP_down.12x11.tiff", height = 9, width = 10, res = 1200, compression = "lzw", units = "cm")

gp=ggplot(dotp)+
  geom_point(aes(x = time, y = term, size = ratio*100, col = adj.p))+
  theme_light() +
  theme(text = element_text(size = 10),
        #legend.position = "none",
        #axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        legend.key.size = unit(0.3,"cm"),
        legend.text = element_text(size=8),
        legend.title = element_text(size=9),
        plot.title = element_text(size = 10,face="bold"),
        #strip.text.x = element_text(face = "italic"),
  )+
  scale_colour_gradientn(colours = c("#4575B4","#FFFFBF"))+
  labs(title = "Top_down_GO:BP",x="",y="")
plot(gp)
dev.off()




# analysis for social behaviour associated gene, contains codes for Fig4a, s.Fig3b, s.Fig4d ---------------------------

source("./code/gene_social_behaviour.R")

# predatory stress (Looming dots) analysis --------------------------------

#ad.res
###LD
#read.LD_tables
#matrix
#1:d120_s.POSvsWT = star_pos_d120 - tu_wt_d120,
#2:d120_LD_s.POSvsWT = star_pos_d120_LD - tu_wt_d120_LD,
#3:d120_s.POSvsd120_LD_s.POS = star_pos_d120_LD - star_pos_d120,
#4:d120_TUvsd120_LD.TU = tu_wt_d120_LD - tu_wt_d120,

source("./code/mRNA-seq_analysis_for_LD.R")


# identification of DEGs, cotains codes for Fig5BCE, Fig6ABC, stable 3,4,5,6,7 --------------------------------------------------


source("./code/LD_tx_anlayses.R")




